File: solutions/haiku-1/haiku/perl/transpose.pl
#!/usr/bin/perl
use strict;
use warnings;

my ($R, $C) = split / /, <>;

my @matrix;
for (1..$R) {
    my $line = <>;
    chomp($line);
    my @row = split / /, $line;
    push @matrix, \@row;
}

# Transpose
for my $j (0..$C-1) {
    my @transposed_row = map { $matrix[$_][$j] } 0..$R-1;
    print join(" ", @transposed_row) . "\n";
}

File: solutions/haiku-1/haiku/rust/transpose.rs
use std::io::{self, BufRead};

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    // Read R and C
    let first_line = lines.next().unwrap().unwrap();
    let parts: Vec<&str> = first_line.split_whitespace().collect();
    let r: usize = parts[0].parse().unwrap();
    let c: usize = parts[1].parse().unwrap();

    // Read the matrix
    let mut matrix: Vec<Vec<i32>> = Vec::new();
    for _ in 0..r {
        let line = lines.next().unwrap().unwrap();
        let row: Vec<i32> = line.split_whitespace()
            .map(|s| s.parse().unwrap())
            .collect();
        matrix.push(row);
    }

    // Transpose and output
    for j in 0..c {
        let mut output = Vec::new();
        for i in 0..r {
            output.push(matrix[i][j].to_string());
        }
        println!("{}", output.join(" "));
    }
}
